conans/client/cmd/export.py

```diff
@@ -447,41 +448,51 @@ def _export_scm(scm_data, origin_folder, scm_sources_folder, output):
 
 
 def export_source(conanfile, origin_folder, destination_source_folder):
-    if isinstance(conanfile.exports_sources, str):
-        conanfile.exports_sources = (conanfile.exports_sources, )
+    if callable(conanfile.exports_sources):
+        conanfile.deps_cpp_info = DepsCppInfo()
+        conanfile.copy = FileCopier([origin_folder], destination_source_folder)
+        conanfile.export_sources()
+    else:
+        if isinstance(conanfile.exports_sources, str):
+            conanfile.exports_sources = (conanfile.exports_sources, )
 
-    included_sources, excluded_sources = _classify_patterns(conanfile.exports_sources)
-    copier = FileCopier([origin_folder], destination_source_folder)
-    for pattern in included_sources:
-        copier(pattern, links=True, excludes=excluded_sources)
-    output = conanfile.output
-    package_output = ScopedOutput("%s exports_sources" % output.scope, output)
-    copier.report(package_output)
+        included_sources, excluded_sources = _classify_patterns(conanfile.exports_sources)
+        copier = FileCopier([origin_folder], destination_source_folder)
+        for pattern in included_sources:
+            copier(pattern, links=True, excludes=excluded_sources)
+        output = conanfile.output
+        package_output = ScopedOutput("%s exports_sources" % output.scope, output)
+        copier.report(package_output)
 
 
 def export_recipe(conanfile, origin_folder, destination_folder):
-    if isinstance(conanfile.exports, str):
-        conanfile.exports = (conanfile.exports, )
+    if callable(conanfile.exports):
+        conanfile.deps_cpp_info = DepsCppInfo()
+        conanfile.copy = FileCopier([origin_folder], destination_folder)
+        conanfile.exports()
+    else:
+        if isinstance(conanfile.exports, str):
+            conanfile.exports = (conanfile.exports, )
 
-    output = conanfile.output
-    package_output = ScopedOutput("%s exports" % output.scope, output)
+        output = conanfile.output
+        package_output = ScopedOutput("%s exports" % output.scope, output)
 
-    if os.path.exists(os.path.join(origin_folder, DATA_YML)):
-        package_output.info("File '{}' found. Exporting it...".format(DATA_YML))
-        tmp = [DATA_YML]
-        if conanfile.exports:
-            tmp.extend(conanfile.exports)  # conanfile.exports could be a tuple (immutable)
-        conanfile.exports = tmp
+        if os.path.exists(os.path.join(origin_folder, DATA_YML)):
```

Comment:
The data_yml needs to be exported always. Probably this should be done irrespective of the attribute/method approach